### ISSUE TITLE
Avoid extraneous goroutines for RPC calls

### ISSUE THREAD

**Issue by akalin-keybase** [keybase/go-framed-msgpack-rpc#41]
Splitting off these notes from https://github.com/keybase/go-framed-msgpack-rpc/issues/26 and https://github.com/keybase/go-framed-msgpack-rpc/pull/34 .

In dispatch.handleCall, we spawn off a goroutine to wait for a cancel or response. This seems extraneous, since we're already blocking in dispatch.Call. But handleCall is called from the callLoop, which serializes incoming RPC calls, replies, and the stop signal, and one RPC call needs to not block subsequent ones.

But if we have a way to send call information to the packetizer, namely the seqnos of calls we expect to hear a reply for, we don't need callLoop at all. dispatch.Call can just look like:

```
func dispatch.Call(ctx, ...) error {
  errCh := d.dispatchMessage(ctx, ...)
  select {
    case <-ctx.Done():
      return ctx.Err()
    case <-d.stopCh:
      return io.EOF
    case err := <-errCh:
      if err != nil {
        return err
      }
  }

  // RPC has been sent.
  // make a new seqno
  replyCh := make(...)
  // a channel to the packetizer, which adds a map entry from seqno to replyCh.
  callInfoCh <- newCallInfo{seqno, ..., replyCh}

  select {
    case <-ctx.Done():
      // another channel to the packetizer, which removes the map entry from seqno to replyCh.
      cancelInfoCh <- cancelInfo{seqno, ...}
      return ctx.Err()
    case <-d.stopCh:
      cancelInfoCh <- cancelInfo{seqno, ...}
      return io.EOF
    case reply := <-replyCh:
    // handle reply
  }
}
```

This is nice, since it's just straight-line code, and there are no extraneous goroutines, and it handles cancellation correctly (i.e., immediately).

But what does the packetHandler loop look like now? It's only a little more complicated. We already have readerLoop serializing all the decode requests. So we can change Decode() and ReadByte() to return an errCh instead of an err, similar to how we're changing Encode() to do the same thing in the PR above. Then packetizeOne will look like this:

```
func (p *packetHandler) packetizeOne() (err error) {
  // get frame
  var n int
  errCh := p.dec.Decode(&n)
  select {
    case callInfo <- callInfoCh:
      // add callInfo.seqno -> callInfo.replyCh mapping
    case cancelInfo <- cancelInfoCh:
      // remove cancelInfo.seqno map entry
    case err := <-errCh:
      if err != nil {
        return err
      }
  }

  // do basically the same thing for get message
}
```

Then the seqno -> replyCh map would be shared with the receiveHandler (no synchronization needed!) and receiveResponse will look like:

```
func (r *receiveHandler) receiveResponse() (err error) {
  ...
  if err = decodeField(r.reader, m, &m.seqno); err != nil {
    return err
  }

  // defer { remove seqno map entry }
  // look up seqno to replyCh mapping, and send the decoded error to it.
}
```

Of course, this isn't urgent at all yet.

**Comment by akalin-keybase** [comment]
cc @maxtaco and @jzila for thoughts.

**Comment by jzila** [comment]
I'm strongly opposed to this particular solution, even if I might agree with the reasons behind it.

This adds more responsibility to the packetizer, and leaks the dispatcher call abstraction back to the packetizer. I'd rather move towards less functionality in the packetizer than more (as it is, its reason for existing is to allow the typing abstraction violation to exist to maintain zero copy).

If the goal here is to redesign RPC to be more clear and have fewer goroutines, I'm all for that. But I'd rather do that independently rather than as part of feature work.

**Comment by jzila** [comment]
Just to be clear, I super love the idea of making calls inline code. That's great.

But if we set out with that goal, then we should do it as a design intention (and incorporate Notify into that too).

**Comment by akalin-keybase** [comment]
Just to be clear, I mean this to be independent from #26 and #34. In other words, a separate issue/design/PR whatever. (Notify is included in here too, but it's pretty straightforward already, and probably won't change much from #34.)

**Comment by jzila** [comment]
In that case I'm on board. But I'd like to avoid leaking the dispatcher abstraction into the packetizer.

**Comment by akalin-keybase** [comment]
Okay, let me see if I can come up with something...

**Comment by maxtaco** [comment]
I don't have strong opinions either way, but if it helps, we can clone and rejigger the encoder/decoder too.  That's not sacred my any means.  But I probably favor the path of least resistance since the other stuff you guys are working on is more important :)

**Comment by akalin-keybase** [comment]
Just to be clear, I don't intend this to be implemented anytime soon -- I just want to get my thoughts down for the next time we circle back to RPC stuff!

Anyway, I think I have a solution to the dispatcher / packetizer problem. Basically, we do still need a top-level event loop, and that's where the stuff to manage the seqno -> replyCh table should go, but right now the top-level event loop is the packetizer loop, and it doesn't have to be.

Basically, the top-level event loop in `transport.run()` should look like

```
...
writerDone := runInBg(t.writerLoop)
packetizerDone := runInBg(t.packetizerLoop)

var err error

outer:
for {
  select {
    // packetizerLoop sends struct { message, err } on t.packetCh.
    case packet := <-t.packetCh:
      if packet.err != nil {
        err = packet.err
        break outer
      }
      err = t.receiver.Receive(packet.msg)
      if err != nil {
        break outer
      }
      case callInfo <- callInfoCh:
        // add callInfo.seqno -> callInfo.replyCh mapping
      case cancelInfo <- cancelInfoCh:
        // remove cancelInfo.seqno map entry
  }
}
```

So in addition to the reader and the writer loop, we have a third packetizer loop, which basically reads from the socket until a message is fully formed, then sends it on t.packetCh. It contains all the logic in packetHandler, as well as some bits from receiveHandler. But now receiver.Receive() takes a fully-formed message instead of reading from the socket directly, so it's a little cleaner.

And so now the top-level loop just basically waits for new messages to arrive on the channel and hands them off to receiver.Receive. It then also manages the seqno -> replyCh mapping as driven by dispatch.Call, so that receiver.Receive can use it to know where to send RPC replies. (Now in dispatch.Call, we need to send on callInfoCh before doing anything else.)

@jzila what do you think?

**Comment by maxtaco** [comment]
Oh, would this design take advantage of framing?  Looks like it might. The packetizer just returns n bytes of raw data, and `t.receiver.Receive` would do the unmarshalling into type data structures? It would introduce another copy, but that might be acceptable

**Comment by akalin-keybase** [comment]
My design would have the packetizer pass a `message` (or a `*message`) to the receiver, i.e. `t.receiver.Receive()` would take a `message`. I think the packetizer can do all the parsing without doing any copies.

**Comment by jzila** [comment]
Fixed in #50
